### PR TITLE
Variable Value linter

### PR DESCRIPTION
**Pull request by justtal**
New variable value linter.

**Which issue, if any, does this resolve?**
#331

### CODE REVIEW

**Comment by shellscape** [comment]
man that linter name rubs me wrong 😆

**Comment by justtal** [comment]
:)

I don't care changing it name at all.

I just didn't understand the need of "never a variable" option.

File: lib/linters/variable_value.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+const util = require('util');
+
+module.exports = {
+    name: 'variableValue',
+    nodeTypes: ['decl'],
+    message: 'The value of [%s] must be a variable.',
+
+    lint: function variableValueLinter (config, node) {
+        if (!config.enabled) {
```
Comment (jwilsson) [inline lib/linters/variable_value.js]: This check isn't necessary, it's checked beforehand in `linter.js`.

File: lib/linters/variable_value.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+            return;
+        }
+
+        if (config.properties.indexOf(node.prop) === -1) {
+            return;
+        }
+
+        if (/^@/i.test(node.value)) {
```
Comment (jwilsson) [inline lib/linters/variable_value.js]: Could you add a comment explaining this line, i.e.
`// It's a variable, so it's allowed`

File: test/specs/linters/variable_value.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    describe('#variableValue()', function () {
+        it('should have the proper node types', function () {
+            const source = 'color: #ABC;';
+
+            return spec.parse(source, function (ast) {
+                expect(spec.linter.nodeTypes).to.include(ast.root.first.type);
+            });
+        });
+
+        it('should ignore if rule disabled', function () {
```
Comment (jwilsson) [inline test/specs/linters/variable_value.js]: Not needed as per above.

File: lib/linters/variable_value.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+const util = require('util');
+
+module.exports = {
+    name: 'variableValue',
+    nodeTypes: ['decl'],
+    message: 'The value of [%s] must be a variable.',
```
Comment (jwilsson) [inline lib/linters/variable_value.js]: We usually put properties etc. in messages in double quotes.

File: test/specs/linters/variable_value.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                const result = spec.linter.lint(options, ast.root.first);
+
+                expect(result).to.be.undefined;
+            });
+        });
+
+        it('should ignore if property not for lint', function () {
+            const source = 'color: #333;';
+            const options = {
+                enabled: true,
```
Comment (jwilsson) [inline test/specs/linters/variable_value.js]: Could you remove all the `enabled: true` since they aren't necessary.

**Comment by jwilsson** [comment]
Thanks for working on this!

I left a few comments, but overall it looks good.

I think that @shellscape wants (and please correct me if I'm wrong) to have a "never" option because we usually allow linters to be really flexible (even if it might seem a bit crazy). And as I said earlier, I'm fine on either name.

**Comment by justtal** [comment]
@jwilsson  Thanks for the review!

So do you want me to change the config to be:

```
{
  "enabled": true,
  "properties": {
    "color": "always",
    "font-size": "never"
  }
}
```

**Comment by jwilsson** [comment]
Yeah, that looks like it would work.

@shellscape What do you think? And do you have a better explanation for the `never` option?

**Comment by shellscape** [comment]
yeah in the spirit of forward-thinking and flexibility, I could foresee people wanting to designate that certain properties never have a variable used for the value. I can't imagine why, but we have other linters that enable stranger things. 

I would suggest the following setup, to reduce the amount of config that would have to be written for projects that want to specify a litany of options for this linter:

```js
{
  "enabled": true,
  "always": [ "color", "background", "font-family", ... ],
  "never": [ "font-size" ]
}
```

**Comment by justtal** [comment]
Done. Please review.

Will it be in the 3.0 release?

File: lib/linters/variable_value.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+const util = require('util');
+
+module.exports = {
+    name: 'variableValue',
+    nodeTypes: ['decl'],
+    message: '%s is not allowed in "%s" property.',
+
+    lint: function variableValueLinter (config, node) {
+        const reg = /@/i;
```
Comment (jwilsson) [inline lib/linters/variable_value.js]: Perhaps name this something like `reVariable` or `variableRegex`.

File: test/specs/linters/variable_value.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            }];
+
+            return spec.parse(source, function (ast) {
+                const result = spec.linter.lint(options, ast.root.first);
+
+                expect(result).to.deep.equal(expected);
+            });
+        });
+
+        it('should allow non variable if in never list', function () {
```
Comment (jwilsson) [inline test/specs/linters/variable_value.js]: I think this one made it in twice (see line 60).

File: test/specs/linters/variable_value.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            };
+
+            return spec.parse(source, function (ast) {
+                const result = spec.linter.lint(options, ast.root.first);
+
+                expect(result).to.be.undefined;
+            });
+        });
+
+        it('should not allow non variable if in always list', function () {
```
Comment (jwilsson) [inline test/specs/linters/variable_value.js]: Could you also add a test for the opposite, i.e. "should allow variable if in always list".

**Comment by jwilsson** [comment]
Just a few more things. Thanks for working with us!

Unfortunately, it won't make it into `3.0.0`, I think it's big enough as it is. It'll definitely make `3.1.0` though, which I'm hoping we'll get out the door pretty soon afterwards.

**Comment by justtal** [comment]
@jwilsson Thanks for the review. Some stupid mistakes :\

Fixed.

3.1.0 would be great.

**Comment by jwilsson** [comment]
Thanks for contributing! 🎉 

I'm hoping to have a release out later today.

**Pull request merged**